pkg/internaltelemetry/client.go

```diff
@@ -0,0 +1,187 @@
+package internaltelemetry
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	metadatautils "github.com/DataDog/datadog-agent/comp/metadata/host/hostimpl/utils"
+	pb "github.com/DataDog/datadog-agent/pkg/proto/pbgo/trace"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/version"
+)
+
+const (
+	telemetryEndpoint = "/api/v2/apmtelemetry"
+)
+
+// Client defines the interface for a telemetry client
+type Client interface {
+	SendLog(level string, message string)
+	SendTraces(traces *pb.Traces)
+}
+
+type client struct {
+	m         sync.Mutex
+	client    httpClient
+	endpoints []*config.Endpoint
+
+	// we can pre-calculate the host payload structure at init time
+	baseEvent Event
+}
+
+// RequestType defines the request type
+type RequestType string
+
+const (
+	// RequestTypeLogs defines the logs request type
+	RequestTypeLogs RequestType = "logs"
+	// RequestTypeTraces defines the traces request type
+	RequestTypeTraces RequestType = "traces"
+)
+
+// Event defines the event object
+type Event struct {
+	APIVersion  string      `json:"api_version"`
+	RequestType RequestType `json:"request_type"`
+	TracerTime  int64       `json:"tracer_time"` // unix timestamp (in seconds)
+	RuntimeID   string      `json:"runtime_id"`
+	SequenceID  uint64      `json:"seq_id"`
+	DebugFlag   bool        `json:"debug"`
+	Host        Host        `json:"host"`
+	Application Application `json:"application"`
+	Payload     interface{} `json:"payload"`
+}
+
+// Host defines the host object
+type Host struct {
+	Hostname      string `json:"hostname"`
+	OS            string `json:"os"`
+	Arch          string `json:"architecture"`
+	KernelName    string `json:"kernel_name"`
+	KernelRelease string `json:"kernel_release"`
+	KernelVersion string `json:"kernel_version"`
+}
+
+// Application defines the application object
+type Application struct {
+	ServiceName     string `json:"service_name"`
+	LanguageName    string `json:"language_name"`
+	LanguageVersion string `json:"language_version"`
+	TracerVersion   string `json:"tracer_version"`
+}
+
+type TracePayload struct {
+	Traces []pb.Trace `json:"traces"`
+}
+
+// LogPayload defines the log payload object
+type LogPayload struct {
+	Logs []LogMessage `json:"logs"`
+}
+
+// LogMessage defines the log message object
+type LogMessage struct {
+	Message string `json:"message"`
+	Level   string `json:"level"`
+}
+
+var sequenceID atomic.Uint64
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// NewClient creates a new telemetry client
+func NewClient(httpClient httpClient, endpoints []*config.Endpoint, service string) Client {
+	info := metadatautils.GetInformation()
+	return &client{
+		client:    httpClient,
+		endpoints: endpoints,
+		baseEvent: Event{
+			APIVersion: "v2",
+			DebugFlag:  true,
```

Comment:
Could you make that a parameter so we don't forget to turn it off in the release?